sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/GroupByKeyEncodingTest.java

```diff
@@ -0,0 +1,221 @@
+ */
+@RunWith(JUnit4.class)
+public class GroupByKeyEncodingTest {
+
+  @Rule
+  public ExpectedException thrown = ExpectedException.none();
+
+  /**
+   * This is a bogus key class that returns random hash values from {@link #hashCode()} and always
+   * returns {@code false} for {@link #equals(Object)}. The results are correct if Flink does not
+   * use this for hashing/sorting.
+   */
+  @DefaultCoder(AvroCoder.class)
+  static class Key {
+    @Nullable
+    String key1;
+    @Nullable
+    Long key2;
+
+    public Key() {}
+
+    public Key(String s, Long k) {
+      key1 = s;
+      key2 = k;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      return false;
+    }
+
+    @Override
+    public int hashCode() {
+      UUID uuid = UUID.randomUUID();
+      return uuid.hashCode();
+    }
+  }
+
+  @DefaultCoder(AvroCoder.class)
+  static class DumbData {
+    @Nullable
+    String key1;
+    @Nullable
+    Long key2;
+    @Nullable
+    Long value1;
+
+    public DumbData() {
+    }
+  }
+
+  @DefaultCoder(AvroCoder.class)
+  static class Config {
+    int key1;
+    int key2;
+    int perKey;
+    long val;
+
+    public Config() {
+    }
+
+    public Config(int k1, int k2, int pk, long v) {
+      key1 = k1;
+      key2 = k2;
+      perKey = pk;
+      val = v;
+    }
+  }
+
+  static class MakeKey extends SimpleFunction<DumbData, Key> {
+
+    @Override
+    public Key apply(DumbData dd) {
+      return new Key(dd.key1, dd.key2);
+    }
+  }
+
+  static class Generator extends DoFn<Config, DumbData> {
+
+    @Override
+    public void processElement(ProcessContext c) throws Exception {
+      Config cfg = c.element();
+
+      ArrayList<String> bs = new ArrayList<>(cfg.key1);
+      for (int i = 0; i < cfg.key1; i++) {
+        bs.add("hereGoesLongStringID" + i);
+      }
+
+      for (int k = 0; k < cfg.perKey; k++) {
+        for (long j = 0; j < cfg.key2; j++) {
+          for (int i = 0; i < cfg.key1; i++) {
+            DumbData dd = new DumbData();
+            dd.key1 = bs.get(i);
+            dd.key2 = j;
+            dd.value1 = cfg.val;
+            c.output(dd);
+          }
+        }
+      }
+    }
+  }
+
+  static class MergeGbk extends DoFn<KV<Key, Iterable<DumbData>>, String> {
+
+    private static final Logger LOG = LoggerFactory.getLogger(MergeGbk.class);
+
+    private final Aggregator<Long, Long> keyCnt =
+        createAggregator("key count", new Sum.SumLongFn());
+
+    private final Aggregator<Long, Long> itemCnt =
+        createAggregator("item count", new Sum.SumLongFn());
+
+    @Override
+    public void processElement(ProcessContext c) throws Exception {
+      Key key = c.element().getKey();
+      Iterable<DumbData> data = c.element().getValue();
+
+      keyCnt.addValue(1L);
+      long count = 0;
+      for (DumbData val : data) {
+        count++;
+      }
+      itemCnt.addValue(count);
+
+      c.output(key.key1 + "," + key.key2 + "," + count);
```

Comment:
I would prefer to use `Coder.structuralValue()` to perform equality tests instead of strings, and as such use `KV<Object, Long>`